vulntotal/datasources/gitlab.py

```diff
@@ -0,0 +1,182 @@
+
+
+def download_subtree(package_slug: str, speculative_execution=False):
+    url = f"https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/archive/master/gemnasium-db-master.tar.gz?path={package_slug}"
+    response = fetch(url)
+    if os.path.getsize(response.location) > 0:
+        extracted_location = Path(response.location).parent.joinpath(
+            "temp_vulntotal_gitlab_datasource"
+        )
+        with tarfile.open(response.location, "r") as file_obj:
+            file_obj.extractall(extracted_location)
+        os.remove(response.location)
+        return extracted_location
+    if not speculative_execution:
+        logger.error(f"{package_slug} doesn't exist")
+    os.remove(response.location)
+
+
+def clear_download(location):
+    if location:
+        shutil.rmtree(location)
+
+
+def get_casesensitive_slug(path, package_slug):
+    payload = [
+        {
+            "operationName": "getPaginatedTree",
+            "variables": {
+                "projectPath": "gitlab-org/security-products/gemnasium-db",
+                "ref": "master",
+                "path": path,
+                "nextPageCursor": "",
+                "pageSize": 100,
+            },
+            "query": """
+            fragment TreeEntry on Entry { 
+                flatPath 
+            } 
+            query getPaginatedTree($projectPath: ID!, $path: String, $ref: String!, $nextPageCursor: String) { 
+                project(fullPath: $projectPath) { 
+                    repository { 
+                        paginatedTree(path: $path, ref: $ref, after: $nextPageCursor) { 
+                            pageInfo { 
+                            endCursor
+                            startCursor
+                            hasNextPage 
+                            } 
+                        nodes { 
+                            trees { 
+                                nodes { 
+                                    ...TreeEntry 
+                                    } 
+                                } 
+                            } 
+                        } 
+                    } 
+                } 
+            } """,
+        }
+    ]
+    url = "https://gitlab.com/api/graphql"
+    hasnext = True
+
+    while hasnext:
+        response = requests.post(url, json=payload).json()
+        paginated_tree = response[0]["data"]["project"]["repository"]["paginatedTree"]
+
+        for slug in paginated_tree["nodes"][0]["trees"]["nodes"]:
+            if slug["flatPath"].lower() == package_slug.lower():
+                return slug["flatPath"]
+
+            # If the namespace/subfolder contains multiple packages, then progressive transverse through folders tree
+            if package_slug.lower().startswith(slug["flatPath"].lower()):
+                return get_gitlab_style_slug(slug["flatPath"], package_slug)
+
+        payload[0]["variables"]["nextPageCursor"] = paginated_tree["pageInfo"]["endCursor"]
```

Comment:
check `get_item`

Reply:
[https://github.com/nexB/vulnerablecode/pull/883#discussion_r963588542](url)

Reply:
Check my comment above for same.